lib/IRGen/TypeLayoutDumper.cpp

```diff
@@ -0,0 +1,212 @@
+          fixedTI->getFixedAlignment().getValue(),
+          fixedTI->getFixedExtraInhabitantCount(IGM)};
+}
+
+static void addYAMLTypeInfoNode(NominalTypeDecl *NTD,
+                                IRGenModule &IGM,
+                                std::vector<YAMLTypeInfoNode> &Result) {
+  // We only care about public and @usableFromInline declarations.
+  if (NTD->getEffectiveAccess() < AccessLevel::Public)
+    return;
+
+  // We don't care about protocols or classes.
+  if (isa<ProtocolDecl>(NTD) || isa<ClassDecl>(NTD))
+    return;
+
+  assert(isa<StructDecl>(NTD) || isa<EnumDecl>(NTD));
+
+  auto &Opts = IGM.getOptions();
+
+  switch (Opts.TypeInfoFilter) {
+  case IRGenOptions::TypeInfoDumpFilter::All:
+    break;
+  case IRGenOptions::TypeInfoDumpFilter::Resilient:
+    if (!NTD->isFormallyResilient())
+      return;
+    break;
+  case IRGenOptions::TypeInfoDumpFilter::Fragile:
+    if (NTD->isFormallyResilient())
+      return;
+    break;
+  }
+
+  auto *TI = &IGM.getTypeInfoForUnlowered(NTD->getDeclaredTypeInContext());
+  auto *fixedTI = dyn_cast<FixedTypeInfo>(TI);
+  if (!fixedTI)
+    return;
+
+  Result.push_back(createYAMLTypeInfoNode(NTD, IGM, fixedTI));
+}
+
+static YAMLModuleNode createYAMLModuleNode(ModuleDecl *Mod,
+                                           IRGenModule &IGM) {
+  std::vector<NominalTypeDecl *> Decls;
+  NominalTypeWalker Walker(Decls);
+
+  // Collect all nominal types, including nested types.
+  SmallVector<Decl *, 16> TopLevelDecls;
+  Mod->getTopLevelDecls(TopLevelDecls);
+
+  for (auto *D : TopLevelDecls)
+    D->walk(Walker);
+
+  std::vector<YAMLTypeInfoNode> Nodes;
+
+  // Convert each nominal type.
+  for (auto *D : Decls) {
+    if (auto *NTD = dyn_cast<NominalTypeDecl>(D)) {
+      addYAMLTypeInfoNode(NTD, IGM, Nodes);
+    }
+  }
+
+  std::sort(Nodes.begin(), Nodes.end());
+
+  return {Mod->getName().str(), Nodes};
+}
+
+void TypeLayoutDumper::write(ArrayRef<ModuleDecl *> AllModules,
+                             llvm::raw_ostream &os) const {
+  llvm::yaml::Output yout(os);
+
+  // Collect all nominal types, including nested types.
+  for (auto *Mod : AllModules) {
+    auto Node = createYAMLModuleNode(Mod, IGM);
+    yout << Node;
+  }
+}
+
+bool swift::performDumpTypeInfo(IRGenOptions &Opts,
+                                std::unique_ptr<SILModule> SILMod,
+                                llvm::LLVMContext &LLVMContext) {
+  auto &Ctx = SILMod->getASTContext();
+  assert(!Ctx.hadError());
+
+  IRGenerator IRGen(Opts, *SILMod);
+  IRGenModule IGM(IRGen, IRGen.createTargetMachine(), LLVMContext);
+
+  // We want to bypass resilience.
+  CompletelyFragileScope scope(IGM);
+
+  auto *Mod = SILMod->getSwiftModule();
+  SmallVector<Decl *, 16> AllDecls;
+  Mod->getTopLevelDecls(AllDecls);
+
+  SmallVector<ModuleDecl *, 4> AllModules;
+  for (auto *D : AllDecls) {
+    if (auto *ID = dyn_cast<ImportDecl>(D)) {
+      if (auto *M = ID->getModule())
+        AllModules.push_back(M);
```

Comment:
If you actually want to collect all visible modules, use `ModuleDecl::forAllVisibleModules`. I think it would make a lot more sense to just operate on the *current* module, though. What's the benefit of tracking down imports?

Reply:
My idea was that instead of building Foundation in this special mode, you build a dummy module that _imports_ Foundation, etc.

Reply:
Why is that better?

Reply:
Because then you don't have to build Foundation twice. Maybe it doesn't matter, though.